### PR TITLE
WIP: Create WP Plugin

### PR DESCRIPTION
**Pull request by ajitbohra**
This PR is the result of casual discussion with @gziolo on `wp-cli` block scaffolding :)

Adds initial `MVP` for scaffolding WP plugin with a simple block using `starter` created by @youknowriad 

**Demo GIF**
create-wp-plugin

**Immediate Items on list**
- Validation and error handling
- Improve CLI experience for the user
- Mustachejs for templates
- Option to chose bundler `parcel`, `webpack` 

Rough Roadmap/wishlist posted here https://github.com/WordPress/gutenberg/issues/10629

### CODE REVIEW

File: docs/reference/scripts.md [hunk 1]
```diff
@@ -26,7 +26,7 @@ The editor includes a number of packages to enable various pieces of functionali
 | [Element](https://wordpress.org/gutenberg/handbook/packages/packages-element/) | wp-element |Element is, quite simply, an abstraction layer atop [React](https://reactjs.org/) |
 | [Escape Html](https://wordpress.org/gutenberg/handbook/packages/packages-escape-html/) | wp-escape-html | Escape HTML utils |
 | [Hooks](https://wordpress.org/gutenberg/handbook/packages/packages-hooks/) | wp-hooks | A lightweight and efficient EventManager for JavaScript |
-| [Html Entities](https://wordpress.org/gutenberg/handbook/packages/packages-html-entities/) | wo-html-entities | HTML entity utilities for WordPress |
+| [Html Entities](https://wordpress.org/gutenberg/handbook/packages/packages-html-entities/) | wp-html-entities | HTML entity utilities for WordPress |
```
Comment (Soean) [inline docs/reference/scripts.md]: Maybe fron another PR?

Comment (ajitbohra) [reply docs/reference/scripts.md]: oops yes

File: packages/create-plugin/templates/plugin/README.md [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+
+Installing the dependencies also creates or updates a `package-lock.json` file. This file ensures that other developpers working on the repository can install the correct versions of the dependencies by just running `npm install` on their own local repositories.
+
+We're set, we installed the parcel bundler. We can use it to define some scripts to bundle our JavaScript file for developement and production. The `package.json` file allows us to define a `scripts` section where we can add shortcuts to the most used command lines while developping a project. So for instance, instead of having to run `parcel build src/index.js` each time we want to build a production version of our JavaScript files, we can define a `build` script in the `package.json` and just run `npm run build`.
+
+At this point, we define our two scripts: the `start` script we're going to use when developping the plugin to automatically build a development version of the plugin and the `build` script. Try running `npm start` or `npm run build` and you'll notice that a bundle file `index.js` is created in the `dist` folder.
```
Comment (Soean) [inline packages/create-plugin/templates/plugin/README.md]: Typo: developing

File: packages/create-plugin/templates/plugin/README.md [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+
+Similarily to most package dependency managers, we need to provide some basic informations in a config file called [`package.json`](https://github.com/youknowriad/wp-js-plugin-starter/blob/bbd655d2796ea0676ebd2f502664ce10e3c31817/package.json): name, current version, license, keywords... This file can be generated automatically by running `npm run init` and replying to some simple questions about your package.
+
+Once initialized, we can start defining and installing our dependencies (the packages we depend on). So we start by installing the Parcel bundler as a devDependency (a dependency useful when developping our current project but a dependency we don't ship in our production code). So we can just run `npm install parcel-bundler --save-dev`. Running this command updates the `package.json` file and adds the newly installed dependency to the `devDependencies` section of the file.
```
Comment (Soean) [inline packages/create-plugin/templates/plugin/README.md]: Typo: developing

File: packages/create-plugin/templates/plugin/README.md [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+
+- `parcel watch [entryPoint]`: A command taking a JavaScript file as entry point and bundling all the dependencies into a `dist` folder producing an unminified bundle for developement purpose. It also updates the bundle each time change one of the files used in the bundle.
+- `parcel build [entryPoint]`: A command bundling the entryPoint and its dependencies into production-ready files.
+
+But first, we need to install these tools we're going to use. In the JavaScript community, we do so by using [npm](http://npmjs.com), a package dependency manager for JavaScript (if you're more familiar with `PHP`, it's the `composer` of `JavaScript`). (Npm is automatically installed with [NodeJS](https://nodejs.org/en/), so just grab the latest Node.JS binary and install it).
+
+Similarily to most package dependency managers, we need to provide some basic informations in a config file called [`package.json`](https://github.com/youknowriad/wp-js-plugin-starter/blob/bbd655d2796ea0676ebd2f502664ce10e3c31817/package.json): name, current version, license, keywords... This file can be generated automatically by running `npm run init` and replying to some simple questions about your package.
```
Comment (Soean) [inline packages/create-plugin/templates/plugin/README.md]: Typo: informations -> information

File: packages/create-plugin/templates/plugin/wp-js-plugin-starter.php [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+ * Plugin Name: {{pluginName}}
+ * Plugin URI: {{pluginUri}}
+ * Description: {{pluginDescription}}
+ * Version: {{pluginVersion}}
+ * Author: {{pluginAuthor}}
+ *
+ * @package {{pluginSlug}}
+ */
+
+ /**
```
Comment (Soean) [inline packages/create-plugin/templates/plugin/wp-js-plugin-starter.php]: Delete blank before slash

**Review by Soean — COMMENTED** [review]
Really cool PR, this has big potential 🔥
Found just a few small typos

**Comment by ajitbohra** [comment]
@Soean thank you will sync some of the changes to `wp-js-plugin-starter` https://github.com/youknowriad/wp-js-plugin-starter

File: packages/create-plugin/templates/plugin/bin/build-plugin-zip.sh [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+status "Generating build..."
+npm run build
+
+# Remove any existing zip file
+rm -f {{pluginSlug}}.zip
+
+# Generate the plugin zip file
+status "Creating archive..."
+zip -r {{pluginSlug}}.zip \
+	{{pluginSlug}}.php \
```
Comment (gziolo) [inline packages/create-plugin/templates/plugin/bin/build-plugin-zip.sh]: I'm wondering how CSS and other assets are handled by Parcel to ensure they are included in the `dist` folder.

**Review by gziolo — COMMENTED** [review]
Can we start by exposing as much as possible from `@wordpress/scripts` before we get to templating? It feels like we can limit this PR to adding the code which handles scaffolding. That would be awesome :)

File: packages/create-plugin/index.js [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+ * Node Version Info
+ */
+const currentNodeVersion = process.versions.node;
+const semver = currentNodeVersion.split( '.' );
+const major = semver[ 0 ];
+
+/**
+ * Check node version
+ */
+if ( major < 8 ) {
```
[the remark this answers is not in the document]
Comment (ajitbohra) [reply packages/create-plugin/index.js]: have been following the scripts would be good to reuse `check-engines` here

File: packages/create-plugin/templates/plugin/docker-compose.yml [hunk 8]
```diff
@@ -0,0 +1,16 @@
+version: "3.1"
```
[the remark this answers is not in the document]
Comment (ajitbohra) [reply packages/create-plugin/templates/plugin/docker-compose.yml]: scripts/configs/templates all would go to `wp-scripts`

File: packages/create-plugin/templates/plugin/package.json [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+    "wordpress",
+    "plugin",
+    "javascript"
+  ],
+  "author": "{{pluginAuthor}}",
+  "license": "{{pluginLicense}}",
+  "devDependencies": {
+    "@babel/core": "^7.1.0",
+    "@wordpress/babel-preset-default": "^2.1.0",
+    "parcel-bundler": "^1.10.1"
```
[the remark this answers is not in the document]
Comment (ajitbohra) [reply packages/create-plugin/templates/plugin/package.json]: seems liked missed the PR seems good fit here

File: packages/create-plugin/templates/plugin/src/index.js [hunk 10]
```diff
@@ -0,0 +1,16 @@
+const { registerBlockType } = wp.blocks;
```
[the remark this answers is not in the document]
Comment (ajitbohra) [reply packages/create-plugin/templates/plugin/src/index.js]: Yes we can use this it contains inline documentations also

**Comment by ajitbohra** [comment]
@gziolo 

will soon resume work on this and get initial scaffolding ready utilizing `wp-scripts`

**Comment by nosolosw** [comment]
The `@wordpress/scripts` package has received some love lately (including support for ESNext/JSX transpilation). We also have better docs such (JavaScript tutorial FTW!). Considering that, I wonder whether this could be rethought as `scripts scaffold` command instead of a separate package.

I like the idea of having a scafolding command, this would be a great addition for onboarding new people to Gutenberg. What do you think @ajitbohra ?

**Comment by youknowriad** [comment]
I think `npm` might have a special packages that start with `create-*`. I like the separation of concerns between `@wordpress/scripts` and a create package. A create package might need dependencies not really needed aside for generating the initial app.

I feel scaffolding is something you do once, and the scripts commands are something you use daily in your package, making it a clear separation of concerns.

@nosolosw What are the reasons that you'd prefer a command instead of a separate package?

**Comment by nosolosw** [comment]
@youknowriad those are good thoughts I hadn't considered deeply :+1:

**Comment by gziolo** [comment]
> I think npm might have a special packages that start with create-*. I like the separation of concerns between @wordpress/scripts and a create package. A create package might need dependencies not really needed aside for generating the initial app.

Yes, those are great reasons to put it in its own package. For reference, this is how it looks with `create-react-app`:
```sh
$ npm init react-app ./my-react-ap
```
https://docs.npmjs.com/cli/init#examples

**Comment by gziolo** [comment]
@ajitbohra this PR is now a bit out of date. Do you plan to revisit it and help us land a very first iteration of the scaffolding tool based on the feedback provided? An alternative approach would be to close this PR and let other contributors work on it based on your initial exploration.

**Comment by ajitbohra** [comment]
@youknowriad @noisysocks +1 for keeping it as separate package

@gziolo resuming work on this PR to address the feedback and suggestions

**Comment by gziolo** [comment]
> resuming work on this PR to address the feedback and suggestions

It's great to see you back. Feel free to remove `Stale` label when it gets updated :)

**Comment by gziolo** [comment]
There is this great work from @youknowriad which uses some related tooling to automate Gutenberg releases in #15699. It would be great to align as much as possible.

**Comment by paaljoachim** [comment]
Hey @ajitbohra 
Can we get a status update on this PR? What is needed to move the PR onward?

**Comment by gziolo** [comment]
I have this npm package up and running which sort of duplicates the functionality proposed in this PR:

https://www.npmjs.com/package/create-wordpress-block
https://github.com/gziolo/create-wordpress-block

I plan to open PR against Gutenberg after some testing with the community.

Should we close this PR?

**Comment by mcsf** [comment]
This has been labeled Stale since April. Thanks for the efforts, but I'll close this for now.

**Pull request closed without merging**